### PR TITLE
Feat/moderation revamp

### LINKED ISSUE #21: changerole does not enforce role hierarchy
### Description

The `changerole` command in `src/orca_bot/cogs/moderation.py` (lines 113 to 153) only checks that the person running it has the `Manage Roles` permission. It never checks where the target role sits relative to the caller's own highest role

Because the bot is the one calling `user.add_roles(role)`, Discord only validates the bot's role position, not the caller's. The native Discord client normally stops a moderator from handing out roles above their own rank, but this command skips that protection. The result is that anyone with Manage Roles can assign any role that sits below the bot's top role, including roles that grant Administrator

#### Steps to reproduce

1. Give a test user a low role that includes only the Manage Roles permission
2. Have that user run /changerole @themselves @SomeHigherRole, where the higher role is below the bot but above the user
3. The role is granted

#### Impact

Vertical privilege escalation. A low trust member can raise themselves or others to a much higher permission level

### PR DESCRIPTION
**Pull request by jsun-dot**
## Summary

Rewrites the moderation cog from scratch, expanding it from two commands (`/kick`, `/changerole`) to a full set of seven (`/kick`, `/ban`, `/unban`, `/timeout`, `/untimeout`, `/purge`, `/changerole`), all built on shared embed, DM, hierarchy-check, and logging helpers. Also adds `staging` as the integration branch and wires ruff + mypy into `make test` and CI. Closes #21 

## Objective

Bring moderation tooling up to the same quality bar as the rest of the bot — consistent embed responses, optional reason tracking, DM notifications to affected users, and structured audit logging — while closing a privilege-escalation gap in `/changerole`.

## Problem / Issues

- The old cog only supported `/kick` and `/changerole`, using raw string constants and plain-text responses with no reason tracking, DM notifications, or structured logging.
- `/changerole` only checked that the caller had `Manage Roles` — it never checked the target role against the caller's own role position. Since the bot (not the caller) performs `add_roles`, Discord only validated the bot's hierarchy, not the caller's. Any member with `Manage Roles` could assign themselves or others a role above their own rank, including roles granting Administrator.
- `make test` only ran pytest — ruff and mypy existed as separate make targets but weren't part of the standard pre-PR check or CI.
- The `development` branch had drifted far behind `main` and wasn't usable as an integration branch.

## Results

- New commands: `/ban` (optional `delete_days` 0-7, DM), `/unban` (by user ID), `/timeout` / `/untimeout` (DM on both), `/purge` (confirmation UI, optional user filter).
- `/kick` and `/changerole` rewritten on the new helpers; `/changerole` now toggles (add/remove) and enforces strict role-hierarchy checks on the caller — fixing the privilege-escalation issue.
- All commands: embed responses, optional `reason` params, DM to target user, structured logs (actor, target, guild, reason), unified error handler.
- 16 new tests covering hierarchy-check boundary cases and embed helpers (85 total passing).
- `make test` now runs ruff + mypy + pytest; CI mirrors this and triggers on `staging`.
- `development` renamed to `staging` and reset to match `main`, now the integration branch for feature work ahead of `main`.

## Test Plan

- [x] `make test` passes (ruff, mypy, pytest — 85 tests)
- [ ] Manually verify `/kick`, `/ban`, `/unban`, `/timeout`, `/untimeout`, `/purge`, `/changerole` in a test server
- [ ] Verify DM notifications are sent on kick/ban/timeout/untimeout
- [ ] Verify hierarchy checks block actions against equal/higher-ranked members and the server owner

### CODE REVIEW

File: src/orca_bot/cogs/moderation.py [hunk 1]
```diff
@@ -1,156 +1,600 @@
...
+        )
+
+    @kick.error
+    @ban.error
+    @unban.error
+    @timeout.error
+    @untimeout.error
+    @purge.error
+    @changerole.error
+    async def moderation_error(
```
Comment (jsun-dot) [inline src/orca_bot/cogs/moderation.py:578]: blocking: On slash invocation, discord.py wraps the original error (HybridCommandError),
so `isinstance(error, commands.MissingPermissions)` is False here and slash
failures fall through to the generic else branch. Unwrap it first before the isinstance chain at line: 583

File: src/orca_bot/cogs/moderation.py [hunk 2]
```diff
@@ -1,156 +1,600 @@
...
+        if interaction.user != self.ctx.author:
+            await interaction.response.send_message(
+                embed=_error_embed("Only the command author can use these buttons."),
+                ephemeral=True,
+            )
+            return False
+        return True
+
+    @discord.ui.button(label="Confirm", style=discord.ButtonStyle.danger)
+    async def confirm(
```
Comment (jsun-dot) [inline src/orca_bot/cogs/moderation.py:102]: blocking: This view never calls self.stop(), so on_timeout (line 172) still fires ~30s
later and overwrites the result even after confirm/cancel ran. Call
self.stop() at the end of both confirm (~152) and cancel (~170). Also disable
the buttons (or call self.stop()) here at the top of confirm, before the
purge call, to prevent a double-click from triggering a double purge.

File: src/orca_bot/cogs/moderation.py [hunk 3]
```diff
@@ -1,156 +1,600 @@
...
+            ephemeral=True,
+        )
+
+    @commands.hybrid_command(
+        name="changerole",
+        description="Add or remove a role from a member (toggles).",
+    )
+    @commands.has_permissions(manage_roles=True)
+    @commands.guild_only()
     async def changerole(
```
Comment (jsun-dot) [inline src/orca_bot/cogs/moderation.py:521]: blocking: changerole only checks the role's position against the bot's/caller's top
role — it never runs _hierarchy_check against the target member. But
README.md (~74-78) and CHANGELOG.md both state that all moderation commands
block when the target outranks the caller or is the server owner, which isn't
true here. Either add _hierarchy_check(ctx.author, ctx.guild.me, user) for the
target, or correct the docs to carve out changerole as role-based rather than
target-based.

File: src/orca_bot/cogs/moderation.py [hunk 4]
```diff
@@ -1,156 +1,600 @@
...
+            _dm_embed(
+                f"You have been banned from {ctx.guild.name}",
+                f"**Reason:** {reason or 'No reason provided.'}",
+            ),
+        )
+
+        try:
+            await user.ban(
+                reason=reason,
+                delete_message_days=delete_days,
```
Comment (jsun-dot) [inline src/orca_bot/cogs/moderation.py:286]: blocking: delete_message_days is deprecated in discord.py and removed in 3.x. Swap to
delete_message_seconds=delete_days * 86400 — works fine on the current 2.4
pin but breaks silently on the next major bump.

File: src/orca_bot/cogs/moderation.py [hunk 5]
```diff
@@ -1,156 +1,600 @@
...
+                ephemeral=True,
             )
             return
 
-        assert ctx.guild is not None
-
-        guild_member = ctx.guild.me
-        if guild_member is None or not guild_member.guild_permissions.manage_roles:
-            await ctx.send(BOT_ROLE_PERMISSION_MESSAGE)
+        if role >= ctx.author.top_role:
```
Comment (jsun-dot) [inline src/orca_bot/cogs/moderation.py:540]: blocking: This check is the fix for #21 (privilege escalation) and the main reason for
this PR, but it has zero test coverage. Please add a test in
tests/test_moderation.py asserting a caller with Manage Roles cannot
assign/remove a role at or above their own top role.

File: src/orca_bot/cogs/moderation.py [hunk 6]
```diff
@@ -1,156 +1,600 @@
...
 
         try:
-            await user.add_roles(role)
+            if role in user.roles:
+                await user.remove_roles(role, reason=f"Role toggled by {ctx.author}")
+                action = "removed from"
+            else:
+                await user.add_roles(role, reason=f"Role toggled by {ctx.author}")
+                action = "added to"
+        except discord.Forbidden:
```
Comment (jsun-dot) [inline src/orca_bot/cogs/moderation.py:556]: non-blocking: Managed/integration roles, @everyone, and booster roles raise
discord.HTTPException (400), not Forbidden, and currently fall through to the
generic "An error occurred" message. Add discord.HTTPException to this
except, or pre-check role.is_assignable() / not role.is_default().

File: src/orca_bot/cogs/moderation.py [hunk 7]
```diff
@@ -1,156 +1,600 @@
...
+    if target == actor.guild.owner:
+        return "You cannot action the server owner."
+    if actor != actor.guild.owner and target.top_role >= actor.top_role:
+        return "You cannot action someone with an equal or higher role than yours."
+    if target.top_role >= bot.top_role:
+        return "My highest role must be above the target's role to perform this action."
+    return None
+
+
+def _log_action(
```
Comment (jsun-dot) [inline src/orca_bot/cogs/moderation.py:59]: non-blocking: Logging ctx.author / target as strings (usernames) rather than IDs makes this
a weak audit trail — usernames can change. Consider also logging actor.id and
the target's ID.

File: src/orca_bot/cogs/moderation.py [hunk 8]
```diff
@@ -1,156 +1,600 @@
...
+            await ctx.send(
+                embed=_error_embed("You cannot ban yourself."), ephemeral=True
+            )
+            return
+
+        if err := _hierarchy_check(ctx.author, ctx.guild.me, user):
+            await ctx.send(embed=_error_embed(err), ephemeral=True)
+            return
+
+        if not 0 <= delete_days <= 7:
```
Comment (jsun-dot) [inline src/orca_bot/cogs/moderation.py:268]: non-blocking: This validation runs after _hierarchy_check (~264).
Move it above so obviously-invalid input is rejected first.

File: src/orca_bot/cogs/moderation.py [hunk 9]
```diff
@@ -1,156 +1,600 @@
...
+        )
+
+    @commands.hybrid_command(name="ban", description="Ban a member from the server.")
+    @commands.has_permissions(ban_members=True)
+    @commands.guild_only()
+    async def ban(
+        self,
+        ctx: commands.Context,
+        user: discord.Member,
+        delete_days: int = 0,
```
Comment (jsun-dot) [inline src/orca_bot/cogs/moderation.py:250]: non-blocking: As a positional int before *reason, the prefix form `ban @user spam toxic`
will try (and fail) to parse "spam" as delete_days instead of treating it as
part of the reason. Slash users are unaffected since delete_days is a
separate field — worth a doc note for prefix users, or make this keyword-only.

File: src/orca_bot/cogs/moderation.py [hunk 10]
```diff
@@ -1,156 +1,600 @@
...
+        ctx: commands.Context,
+        user: discord.Member,
+        *,
+        reason: str | None = None,
+    ) -> None:
+        """Kick a member from the guild."""
+
+        await ctx.defer(ephemeral=True)
+
+        if user == ctx.author:
```
Comment (jsun-dot) [inline src/orca_bot/cogs/moderation.py:206]: non-blocking: _hierarchy_check below already blocks acting on a target with an equal-or-
higher role, which covers user == ctx.author. This explicit self-check is
redundant — keep it only for the friendlier "You cannot kick/ban yourself"
message, otherwise it can be removed. (Same applies to the ban self-check at
~258.)

File: src/orca_bot/cogs/moderation.py [hunk 11]
```diff
@@ -1,156 +1,600 @@
...
+        _button: discord.ui.Button,
+    ) -> None:
+        await interaction.response.defer()
+        limit = None if self.amount == "All" else int(self.amount)
+
+        def check(m: discord.Message) -> bool:
+            return self.target_user is None or m.author == self.target_user
+
+        try:
+            deleted = await interaction.channel.purge(limit=limit, check=check)
```
Comment (jsun-dot) [inline src/orca_bot/cogs/moderation.py:114]: non-blocking: interaction.channel is typed Optional and not every channel type exposes
.purge() (e.g. forum channels, DMs). Worth running ruff/mypy/pytest locally on
this branch rather than trusting the CI green check — this line looks like a
likely mypy flag.

**Review by jsun-dot — COMMENTED** [review]
couple of blockers here that need to be addressed. Other issues are non blocking but should also be addressed for clean code.
